### ISSUE TITLE
NeedHelp

### ISSUE THREAD

**Issue by cailuming** [turanszkij/WickedEngine#455]
Hi there, if I have a brick, and export it to wiscene, then how can I use it like a prefab, I have tried like this:

- 1.Load the wiscene.
- 2.Create several objectComponent
- 3.Find the meshID by FindEntityByName.
- 4.Assign the meshID to each Object->meshID

Here is the sample code:
 ```
        Entity brick = wi::scene::GetScene().Entity_CreateObject("Brick _"+GUID);
        ObjectComponent	*objectComponent = wi::scene::GetScene().objects.GetComponent(brick );
	objectComponent->meshID = GetScene().Entity_FindByName("Castle_Brick_Mesh");
	transformComponent = GetScene().transforms.GetComponent(brick );
        transformComponent->Translate(XMFLOAT3(1,2,3));
```
As a result ,I couldn't get what I want

**Comment by turanszkij** [comment]
You could use `scene.Entity_Duplicate()` for example. But there are other options, too, for example in the editor's terrain generator, I am loading 3 scenes (tree, rock, bush), then reuse their meshes manually, you could check that out. Or an other option, when you loaded a scene from file into a `wi::Archive`, you can serialize that into a `Scene` and use `scene.Merge()` to move your loaded scene into the main scene.

**Comment by cailuming** [comment]
![image](https://user-images.githubusercontent.com/19427259/172175641-7908f160-47bf-4254-ac1e-4206604eccf2.png)
```
Entity tempEntity = GetScene().Entity_FindByName("FortressGate.glb");

gate = GetScene().Entity_Duplicate(tempEntity);
GetScene().names.GetComponent(gate)->name = "Castle_Gate_" + GUID;
 
objectComponent = GetScene().objects.GetComponent(gate);
transformComponent = GetScene().transforms.GetComponent(gate);
```
it seems that when I copy the "FortressGate" which is an root node for both roof and building more than three times, I will crash, but two times will not.
![image](https://user-images.githubusercontent.com/19427259/172178651-6d20b578-defa-41a7-9202-2e0b6eba4b47.png)
the crash shoot
![image](https://user-images.githubusercontent.com/19427259/172179071-f8618a1e-c75d-4490-848f-b0339ca47bb8.png)
it's suspected that when the wiscene file has more than one mesh as children , it will crash , but works well if the scene file has only one mesh.
Here are the test scene files
[single.zip](https://github.com/turanszkij/WickedEngine/files/8845084/single.zip)
[multi.zip](https://github.com/turanszkij/WickedEngine/files/8845085/multi.zip)

**Comment by turanszkij** [comment]
This seems to be a bug when multithreaded jobs being alive while recursive serializations are called, this should fix it: https://github.com/turanszkij/WickedEngine/commit/9cf40b88e64d7d6849262b86ef54e61cc6a98258

But I am not sure I reproduced your exact crash, my repro was duplicating the root object in multi.zip inside the Editor. Let me know if this doesn't fix your problem.

**Comment by cailuming** [comment]
Almost, do not crash now though, but weird
below is for 8 instances for "wall" and "wall corner" each 
![image](https://user-images.githubusercontent.com/19427259/172389137-1776f593-ee8c-4234-b5dc-317a94355fdd.png)
and the code snippet 
![image](https://user-images.githubusercontent.com/19427259/172389800-be7cd426-62cf-41e9-89f6-29d8df80a605.png)
where the corner code 
![image](https://user-images.githubusercontent.com/19427259/172390027-3df94edc-1993-4f10-bdbf-88cc528bd04a.png)
and wall code
![image](https://user-images.githubusercontent.com/19427259/172390152-7a59f23e-d083-4f3c-9461-7562489cd676.png)

However, when I just spawn only one of either , It seems work well
for wall
![image](https://user-images.githubusercontent.com/19427259/172391530-4e82b255-ea8d-49fc-a05e-89ac2ca3acd9.png)
![image](https://user-images.githubusercontent.com/19427259/172390874-cf52c9c3-b7cf-4602-b642-80af15144035.png)

for corner
![image](https://user-images.githubusercontent.com/19427259/172391340-917fdfd1-e940-45a6-aa03-026a73a0b22c.png)
![image](https://user-images.githubusercontent.com/19427259/172391077-28202842-53f8-41f2-82b8-5df7643fe37c.png)
for gate
![image](https://user-images.githubusercontent.com/19427259/172391885-48d8c13c-1ce9-4620-945d-f6595b1b985b.png)
![image](https://user-images.githubusercontent.com/19427259/172391806-3dcfd71d-f6a9-4602-add5-8c3ef1ce012c.png)

By the way, all the constructor functions of the classes are empty but alloc a memory 
![image](https://user-images.githubusercontent.com/19427259/172392337-b78adb6c-bf2e-4f69-bd10-eb23e2cd8d13.png)

In conclusion, I think maybe some mesh or component is overlapped by other something. Or I missed something? 
Here is the test file
[Fortress.zip](https://github.com/turanszkij/WickedEngine/files/8853591/Fortress.zip)

**Comment by cailuming** [comment]
By the way, Entity_Duplicate will not batch the Instance, I just draw 10 objects, shared totally 4 meshes, But turns out 10 drawcalls 
![image](https://user-images.githubusercontent.com/19427259/172629257-a3a3fc97-a2be-4df4-9916-395d366ee155.png)

and I guess the above question with some instances lost is a matter of transform data lost, I checked the RenderDoc, the mesh instance does exist but not correct position with zero point.

**Comment by cailuming** [comment]
[GameProject.zip](https://github.com/turanszkij/WickedEngine/files/8870596/GameProject.zip)
Here is the whole project code, I'm trying to find  where the transforms  of objects are used  in wiRenderer, But nothing valuable information found

**Comment by cailuming** [comment]
[Assets.zip](https://github.com/turanszkij/WickedEngine/files/8870636/Assets.zip)
And the assets, just one scene file

**Comment by cailuming** [comment]
Oh, I got it , thanks , Regarding the Instancing, I have not found DrawInstanceIndirect something,  could it be bad in this case?

**Comment by cailuming** [comment]
Ok, thanks always

**Comment by cailuming** [comment]
Hello, I found the wiScene::Entity_Duplicate(rootEntity) will make the rootEntity parent for all the same duplicated entity including their children. for example,
entityA = wiScene::Entity_Duplicate(rootEntity);
entityB = wiScene::Entity_Duplicate(rootEntity);
entityC = wiScene::Entity_Duplicate(rootEntity);

Well, however, when I call wiScene::Entity_Remove(rootEntity) , entityA,entityB and entityC are also remove accordingly. this is not I expected.

**Comment by turanszkij** [comment]
This is not expected, I have just tested this in a lua script:
```
root = LoadModel("../models/teapot.wiscene")
a = scene.Entity_Duplicate(root)
b = scene.Entity_Duplicate(root)
scene.Entity_Remove(root)
```
In this case, `a` and `b` entities and all components and children still remained in the scene as needed.
By the way, there is a second parameter to `Entity_Remove()`, which says if the removal will be recursive or not. Recursive means that it will delete children as well, this is default behaviour. You could try setting `recursive = false` and see if this does what you want.

**Comment by cailuming** [comment]
```
Entity gate=LoadModel("Assets/Models/FortressGate.wiscene",XMMatrixIdentity(),true);
Entity	a = GetScene().Entity_Duplicate(gate);
Entity	b = GetScene().Entity_Duplicate(gate);
GetScene().Entity_Remove(gate);
```
[FortressGate.zip](https://github.com/turanszkij/WickedEngine/files/8883656/FortressGate.zip)
Teapot certainly no problem, because The teapot file has no root node, you can test with the above file and code.

By the way, what's the size of the terrain, for example, I need to navigate my role on the mini map, so I need to know the terrain size for certain.

**Comment by turanszkij** [comment]
Ok, I see the issue, but it is not really something that's fixable on engine side (I think). The problem is that the mesh is child of the root entity. Root gets duplicated, so the mesh gets duplicated as well, but the object still uses the mesh from the root. Entity references within components can't change really without some manual code because the duplicate function has no idea about what components are, it just uses serialization. I'll think about this problem, maybe there is a better solution, but you can for example make the mesh not be a child of the gate entity.

> By the way, what's the size of the terrain, for example, I need to navigate my role on the mini map, so I need to know the terrain size for certain.

Size of the terrain is infinite, but it generates mesh chunks only around the camera. The distance can be tweaked with the `generation distance` slider.

**Comment by cailuming** [comment]
Yes, definitely I can make the mesh to be not a child of the root entity which is intended to be removed. thanks to tell me what's going on.
As for the terrain, I am not able to find the terrain brush something, with which I can draw the height or flatten or rise the ground,or paint and blend the terrain background texture

**Comment by turanszkij** [comment]
> As for the terrain, I am not able to find the terrain brush something, with which I can draw the height or flatten or rise the ground,or paint and blend the terrain background texture

This is not implemented yet. The paint tool can be used to paint textures or sculpting like for regular meshes, but there is no terrain specific paint modes yet. I plan to do this, written down already: #421

**Comment by cailuming** [comment]
Good morning，I found the case that if I create a empty Object, say  
Entity object = GetScene().Entity_CreateObject("object_name");
Next,I call wi::renderer::SetVoxelRadianceEnabled(true); 
It will crash with the following
![image](https://user-images.githubusercontent.com/19427259/173215851-8e2c966d-c9e4-41e3-aee7-c0d7e19c08a4.png)
off course I can set the object to renderable(false) to avoid this. I'm not sure whether it's a issue though

**Comment by turanszkij** [comment]
I made a fix for this: https://github.com/turanszkij/WickedEngine/commit/a9bb120506800c1aad56d1fbdb1f700e81f2e631
The problem was that object AABB remained invalid (negative) and it incorrectly returned intersections, so it was included in the voxel GI rendering.

**Comment by cailuming** [comment]
Failed to compile the current master branch ,
![SharedScreenshot](https://user-images.githubusercontent.com/19427259/188798954-bfd0bb39-f940-455a-868e-b8c92a06f8b3.jpg)
and it would be better to use english for the comments accordingly. 
the error stack is 

![SharedScreenshot1](https://user-images.githubusercontent.com/19427259/188799286-827683bd-7be6-44a6-8919-079823edc904.jpg)

**Comment by turanszkij** [comment]
Sorry about that, I removed the comment.

**Comment by cailuming** [comment]
It seems the CameraComponent has no method to set view matrix directly. It's necessary in the case that the camera rotates around  and keep looking at some point. I found It's hard to do this by only using CameraComponent::TransformCamera(...).
So I have constructed a custom view matrix, and need to replace the original view matrix.

**Comment by turanszkij** [comment]
The camera view matrix will be computed inside CameraComponent::UpdateCamera() based on CameraComponent's Eye, At, Up values. You can directly modify those values.
- Eye : camera position
- At: lookAt direction (normalized)
- Up: Up direction

**Comment by cailuming** [comment]
![1663581099379](https://user-images.githubusercontent.com/19427259/190992720-95cd5667-ed1e-4cd0-92a0-f225955a3aae.png)
Crash on clicking SaveHeightMap with win+D right away, both Vulkan and D3D.

**Comment by turanszkij** [comment]
The Win + D issue is fixed in this commit: https://github.com/turanszkij/WickedEngine/commit/d832ee167fa1c8cbb8498151cb413d09a9ff2102

**Comment by cailuming** [comment]
![1663846923034](https://user-images.githubusercontent.com/19427259/191737981-84d23eaf-df19-4bb2-b3dc-9b338226b8b6.png)
Crash when drag the render scale slider on both Vulkan and D3D

**Comment by turanszkij** [comment]
Doesn't repro for me. It might be that some graphics option causes this. You can try resetting default graphics options (delete Editor/config.ini or reset from git). Or you can attach your config.ini here and I will try repro it on my side using it.

**Comment by cailuming** [comment]
[config.zip](https://github.com/turanszkij/WickedEngine/files/9640800/config.zip)
Yes, the crash is caused by Editor config, it works well when I delete my config.ini , so here is it.
